src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,103 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int currentCapacity;
+    private T[] listArray;
+    private int size = 0;
+
+    @SuppressWarnings("unchecked")
+    public ArrayList() {
+        this.listArray = (T[]) new Object[DEFAULT_CAPACITY];
+        currentCapacity = DEFAULT_CAPACITY;
+    }
+
     @Override
     public void add(T value) {
-
+        add(value, size);
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public void add(T value, int index) {
-
+        if (index >= 0 && index <= size) {
+            if (size == currentCapacity) {
+                currentCapacity = currentCapacity + (currentCapacity >> 1);
+                T[] temp = (T[]) new Object[currentCapacity];
+                System.arraycopy(listArray, 0, temp, 0, index);
+                System.arraycopy(listArray, index, temp, index + 1, size - index);
+                listArray = temp;
```

Comment:
move lines 27, 28, 29, 31 to separate method
change line 30 so it will copy array in itself :
System.arraycopy(listArray, index, listArray, index + 1, size - index);

Reply:
1. Lines 27 - 31 are unique, they are not repeated anywhere - therefore they do not need to be moved to a separate method.
My opinion: an additional method always requires copying its parameters, and when used in bulk, this will lead to a performance drop.

2. The proposal to change the 30th line is not correct at all, the program will not work this way. The third parameter is the array to which the data is copied, in our partner it should be a temporary array.